### PR TITLE
feat: Pass connectionID and invitationID in events

### LINKED ISSUE #403: Pass connectionID and  invitationID in events
Event need to pass connectionID and eventID to consumers.

### PR DESCRIPTION
**Pull request by fqutishat**
- Pass connectionID and invitationID from did exchange service

closes #403

### CODE REVIEW

File: pkg/didcomm/protocol/didexchange/service.go [hunk 1]
```diff
@@ -46,6 +46,10 @@ const (
 	ConnectionAck = DIDExchangeSpec + "ack"
 	// DIDExchangeServiceType is the service type to be used in DID document
 	DIDExchangeServiceType = "did-communication"
+	// ConnectionID connection id
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: suggest describing the field. Similar for L51

File: pkg/didcomm/protocol/didexchange/service.go [hunk 2]
```diff
@@ -204,58 +211,52 @@ func (s *Service) handle(msg *message) error {
...
-			return fmt.Errorf("failed to persist state %s %w", followup.Name(), err)
+		logger.Infof("finish execute next state: %s", next.Name())
+		if err = s.update(msg.ThreadID, next); err != nil {
+			return fmt.Errorf("failed to persist state %s %w", next.Name(), err)
 		}
-		logger.Infof("persisted the connection using %s and updating the state to %s", msg.ThreadID, followup.Name())
-		err = action()
-		if err != nil {
-			return fmt.Errorf("failed to execute state action %s %w", followup.Name(), err)
+		logger.Infof("persisted the connection using %s and updating the state to %s",
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: seems the state is already updated at this point?

File: pkg/didcomm/protocol/didexchange/service.go [hunk 3]
```diff
@@ -204,58 +211,52 @@ func (s *Service) handle(msg *message) error {
...
+			msg.ThreadID, next.Name())
+		if err := action(); err != nil {
+			return fmt.Errorf("failed to execute state action %s %w", next.Name(), err)
 		}
-		logger.Infof("finish execute state action: %s", followup.Name())
+		logger.Infof("finish execute state action: %s", next.Name())
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: there are too many individual log messages.

File: pkg/didcomm/protocol/didexchange/service.go [hunk 4]
```diff
@@ -204,58 +211,52 @@ func (s *Service) handle(msg *message) error {
...
 
-		logger.Infof("send post event for state %s", followup.Name())
-		s.sendMsgEvents(&dispatcher.StateMsg{Type: dispatcher.PostState, Msg: msg.Msg, StateID: followup.Name()})
+		logger.Infof("send post event for state %s", next.Name())
+		// TODO change from thread id to connection id
+		// TODO pass invitation id
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: I'm confused by the TODOs. Are the invitation and connection IDs not being passed?

File: pkg/didcomm/protocol/didexchange/service.go [hunk 5]
```diff
@@ -128,7 +132,10 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	}
 	// trigger message events
 	logger.Infof("send pre event for state %s", next.Name())
-	s.sendMsgEvents(&dispatcher.StateMsg{Type: dispatcher.PreState, Msg: msg, StateID: next.Name()})
+	// TODO change from thread id to connection id
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: Issue number?

File: pkg/didcomm/protocol/didexchange/service.go [hunk 6]
```diff
@@ -204,58 +211,52 @@ func (s *Service) handle(msg *message) error {
...
+		s.sendMsgEvents(&dispatcher.StateMsg{
+			Type: dispatcher.PostState, Msg: msg.Msg, StateID: next.Name(),
+			Properties: s.createEventProperties(msg.ThreadID, "")})
+		next = followup
 	}
-	logger.Infof("--exited internal handle function--")
 	return nil
 }
 
+func (s *Service) createEventProperties(connectionID, invitationID string) map[string]interface{} { //nolint: unparam
```
Comment (llorllale) [inline pkg/didcomm/protocol/didexchange/service.go]: this method's parameter is named `connectionID` yet I see the `threadID` passed in in all invocations.

File: pkg/didcomm/dispatcher/api.go [hunk 7]
```diff
@@ -47,6 +47,8 @@ type StateMsg struct {
 	Type    StateMsgType
 	StateID string
 	Msg     DIDCommMsg
+	// Properties contains specific keys from different protocols
+	Properties map[string]interface{}
```
Comment (llorllale) [inline pkg/didcomm/dispatcher/api.go]: Suggest we rename this to "Context". The godoc is misleading - in the context of Aries, "keys" usually means "cryptographic keys". Suggest we make this a bit more clear.

File: pkg/didcomm/protocol/didexchange/service.go (hunk 1 above)
Comment (llorllale) [reply pkg/didcomm/protocol/didexchange/service.go]: do these fields need to be exported?

Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: @llorllale https://github.com/hyperledger/aries-framework-go/pull/422/files#diff-427522b353a28b1ec350c29c79d387b0R27

Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: @troyronda done

File: pkg/didcomm/protocol/didexchange/service.go (hunk 2 above)
Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: done

File: pkg/didcomm/dispatcher/api.go (hunk 7 above)
Comment (troyronda) [reply pkg/didcomm/dispatcher/api.go]: +1 on using a term other than keys for the reason given by @llorllale .

I'm not sure, though, that we should use the term context in yet another way :).

File: pkg/didcomm/protocol/didexchange/service.go (hunk 3 above)
Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: We will change all them to debug after we complete the 0.1.0 demo

File: pkg/didcomm/protocol/didexchange/service.go (hunk 4 above)
Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: we suppose to pass the connection id and invitation id but issue #397 not read. So to unblock myself I pass thread id as connection id

File: pkg/didcomm/protocol/didexchange/service.go (hunk 5 above)
Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: done

File: pkg/didcomm/protocol/didexchange/service.go (hunk 6 above)
Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: we suppose to pass the connection id and invitation id but issue #397 not read. So to unblock myself I pass thread id as connection id

File: pkg/didcomm/protocol/didexchange/service.go (hunk 3 above)
Comment (troyronda) [reply pkg/didcomm/protocol/didexchange/service.go]: Even when you change to debug, you should have two logger statements in a row ;).

File: pkg/didcomm/dispatcher/api.go (hunk 7 above)
Comment (fqutishat) [reply pkg/didcomm/dispatcher/api.go]: I change the comments

File: pkg/didcomm/protocol/didexchange/service.go [hunk 8]
```diff
@@ -204,58 +211,51 @@ func (s *Service) handle(msg *message) error {
...
+		logger.Infof("finish execute state action: %s", next.Name())
 
-		logger.Infof("send post event for state %s", followup.Name())
-		s.sendMsgEvents(&dispatcher.StateMsg{Type: dispatcher.PostState, Msg: msg.Msg, StateID: followup.Name()})
+		// TODO change from thread id to connection id #397
+		// TODO pass invitation id #397
+		s.sendMsgEvents(&dispatcher.StateMsg{
+			Type: dispatcher.PostState, Msg: msg.Msg, StateID: next.Name(),
+			Properties: s.createEventProperties(msg.ThreadID, "")})
+		logger.Infof("send post event for state %s", next.Name())
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/service.go]: Sent?

File: pkg/didcomm/protocol/didexchange/service.go (hunk 3 above)
Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: @troyronda I change it

File: pkg/didcomm/protocol/didexchange/service.go (hunk 8 above)
Comment (fqutishat) [reply pkg/didcomm/protocol/didexchange/service.go]: done

File: pkg/didcomm/dispatcher/api.go (hunk 7 above)
Comment (llorllale) [reply pkg/didcomm/dispatcher/api.go]: > I'm not sure, though, that we should use the term context in yet another way :).

I was expecting this counterpoint, but this is really what it is. :)

As a matter of fact, we might want to consider using the actual `context.Context` in the future. The timeout/deadline functionality would align nicely with the `~timing` decorator, and with protocol-specific K-Vs stuffed into it, just like in this case.

**Pull request merged**